music/src/core/player.ts

```diff
@@ -328,9 +328,89 @@ export class SoundFontPlayer extends BasePlayer {
 }
 
 /**
- * A `NoteSequence` player based on Tone.js that includes a click track.
+ * A base class for providing arbitrary callbacks for each note played.
  */
-export class PlayerWithClick extends Player {
+export class BasePlayerCallback {
+
+  /* The function should be overridden, it will be called for each time/note
+   * pair in a sequence being played.
+   */
+  run(t: number, n: NoteSequence.INote) {
+    // Do nothing.
+  }
+
+    /* The function should be overridden, it will be called when a sequence is
+     * stopped.
+     */
+  stop() {
+    // Do nothing.
+  }
+}
+
+/**
+ * A `NoteSequence` player based on Tone.js that allows arbitrary callbacks
+ * when playing each note.
+ */
+export class PlayerWithCallback extends Player {
+  private playClick: boolean;
+  private callbackObject: BasePlayerCallback;
+
+  constructor(playClick?: boolean, callbackObject?: BasePlayerCallback) {
+    super();
+    if (!playClick) {
+      this.playClick = true;
+    } else {
+      this.playClick = playClick;
+    }
+    if (!callbackObject) {
+      this.callbackObject = new BasePlayerCallback();
```

Comment:
just leave this as undefined if it's not provided and use conditionals below

Reply:
done